### DISCUSSION TITLE
VRP Dynamic drop-off window

### DISCUSSION THREAD

**Discussion by odashevsky** [google/or-tools#3278] [Routing (and legacy CP) questions]
Hello,

We're solving a VRCP problem with pickup and drop-off and time windows. We're having an issue where we need to define a drop-off window for its corresponding pickup-up depending on when the pick-up occurred. 

Example: I have two points A and B. Window for pickup @ A is X and Y (say, 1 hour interval). I need to defined the drop-off window @ B in such a way that the load on the track sits for no longer then 30 minutes. Meaning if load is picked up at time T (which is between X and Y), drop off window is between `T + travel time from A to B` and `T + travel time from A to B + 30 minutes`. 
Any idea how to define such a constraint? Thank you.

**Comment by Mizux** [comment]
if you have a time dimension you can add some constraint instead of a range
something like this:
```python
solver = routing.solver() # get the underlying Cp solver
index_a = manager.NodeToIndex(A)
index_b = manager.NodeToIndex(B)
time_dimension = routing.GetDimensionOrDie('time') # I suppose you named your routing dimension 'time'
# Actually you could avoid this constraint or just time A < time B i.e. pickup before delivery no min time
solver.Add(time_dimension.CumulVar(B) >= time_dimension.CumulVar(A) + time_transit[A][B])
# limit to 30min delay max (hard constraint)
solver.Add(time_dimension.CumulVar(B) <= time_dimension.CumulVar(A) + time_transit[A][B] + 30min)
```

note: 30min need to be converted in the unit you use a time unit (seconds, minutes, 1 unit = 7min etc...)

**Reply by odashevsky** [reply]
Hi Mizux.

I tried that and it hangs the solver. It just sits there producing no solutions. It is allowed to drop trips with penalties. Any idea what I should be looking at?

Edit: if extend the "extra" 30 minutes by another 60 minutes, it does find solutions. But why is it handing there w/out any solutions found even though it is allowed to drop locations?

Thanks.

**Reply by Mizux** [reply]
The solver must fulfill all constraints so if solver only drop one node the constraint is broken, I guess it need to drop both nodes A and B at a time.

* one test would be to multiply by `routing.activeVar(A) * routing.activeVar(B)` coef.
```py
active_a_b = routing.activeVar(A) * routing.activeVar(B)
solver.Add(active_a_b * time_dimension.CumulVar(B) >= active_a_b * (time_dimension.CumulVar(A) + time_transit[A][B]))
```

**Reply by jmarca** [reply]
I second the recommendation to multiply by the active var (which is zero or one).  I've noticed that the solver does *not* zero out the cumulvars of nodes that are dropped.  

So if A and B are in a solution, and then later dropped from the solution, the solver will leave the old CumulVar(A) and CumulVar(B) as is.  This leads to unexpected behavior, like a model never converging, if you forget to account for the activevar part.

I often make intermediate variables

```
activeLoadA = time_dimension.CumulVar(A) *  routing.ActiveVar(A)
```
etc

Also, in a pickup and dropoff problem, I *believe* it should always be true that `ActiveVar(A) == ActiveVar(B)`

**Reply by Mizux** [reply]
IIRC if nodes a and b have been flagged as a P&D pairs, solver take extra care of them at least during the search of the initial solution.
Solver should manage to add both at a time and I  guess guided local search operator should do the same the equality `ActiveVar(A) == ActiveVar(B)` should be respected BUT as jmarca said they can have undefined value when dropped and solver still check the constraint....

**Reply by odashevsky** [reply]
Hello,

I would like to revive this thread. I solved the problem using other approaches, but would like to come back and review the approach proposed above. I tried to use

`solver.Add(time_dimension.CumulVar(B) <= time_dimension.CumulVar(A) + time_transit[A][B] + 30min)`

as suggested previously, but the solution produces no results. Adding the 

```
activeLoadA = time_dimension.CumulVar(A) *  routing.ActiveVar(A)
solver.Add(active_a_b * time_dimension.CumulVar(B) >= active_a_b * (time_dimension.CumulVar(A) + time_transit[A][B]))
```

constrains does not make it find any solutions neither. However, I noticed if I change from

`search_parameters.first_solution_strategy = routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC)`

to 

`search_parameters.first_solution_strategy = routing_enums_pb2.FirstSolutionStrategy.PARALLEL_CHEAPEST_INSERTION)`

the solver works find and starts finding solutions.
The same was observed in Mezux's sample here:
[](https://gist.github.com/Mizux/50095103bb39e95511be4a1e53cf724f)
https://gist.github.com/Mizux/50095103bb39e95511be4a1e53cf724f

My questions are:
1. Why PATH_CHEAPEST_ARC did not work with the condition `solver.Add(time_dimension.CumulVar(B) <= time_dimension.CumulVar(A) + time_transit[A][B] + 30min)` in place?
2. What's the difference really between PATH_CHEAPEST_ARC and PARALLEL_CHEAPEST_INSERTION? I am hesitant to switch to PARALLEL_CHEAPEST_INSERTION strategy as I am not sure how will it change finding solutions besides the facts that it will start working with the condition added in this example. There're multiple dimension and other complicated constrains in condition in the problem I am solving and not sure how PARALLEL_CHEAPEST_INSERTION will handle those.

Thank you.
